File: 2022/lib/grid.cpp
#include <ostream>
#include <vector>

#include "point.cpp"

namespace aoc {
using namespace std;

template <typename T> class Grid {
public:
  Grid(const vector<string>& lines, bool diagonal = false)
      : diagonal_(diagonal) {
    for (const string& line : lines)
      // FIXME: Conditional logic based on T, e.g. call stoi if type is int.
      grid_.push_back(vector<T>(line.begin(), line.end()));
  }

  T operator[](Point p) const { return grid_[p.y][p.x]; }
  T& operator[](Point p) { return grid_[p.y][p.x]; }

  typename vector<vector<T>>::iterator begin() { return grid_.begin(); }
  typename vector<vector<T>>::iterator end() { return grid_.end(); }
  typename vector<vector<T>>::const_iterator begin() const {
    return grid_.cbegin();
  }
  typename vector<vector<T>>::const_iterator end() const {
    return grid_.cend();
  }

  bool in_bounds(Point p) const {
    return 0 <= p.y && p.y < grid_.size() && 0 <= p.x &&
           p.x < grid_[p.y].size();
  }
  vector<Point> get_neighbors(Point p) const {
    const static vector<Point> adjacents{{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
    const static vector<Point> diagonals{{1, 1}, {1, -1}, {-1, 1}, {-1, -1}};

    vector<Point> neighbors;
    for (Point n : adjacents)
      if (in_bounds(p + n))
        neighbors.push_back(p + n);

    if (diagonal_)
      for (Point n : diagonals)
        if (in_bounds(p + n))
          neighbors.push_back(p + n);

    return neighbors;
  }

  int height() const { return grid_.size(); }
  int width() const { return grid_[0].size(); }

  // TODO: This would be better as an iterator.
  vector<Point> get_points() {
    vector<Point> points;
    for (int y = 0; y < grid_.size(); ++y)
      for (int x = 0; x < grid_[y].size(); ++x)
        points.push_back({x, y});
    return points;
  }

private:
  vector<vector<T>> grid_;
  bool diagonal_ = false;
};

template <typename T> ostream& operator<<(ostream& out, const Grid<T>& g) {
  for (const auto& row : g) {
    for (const auto& x : row)
      out << x;
    out << '\n';
  }
  return out;
}

} // namespace aoc

File: lib/grid.py
from typing import Optional


class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __hash__(self) -> int:
        return hash((self.x, self.y))

    def __add__(self, other: "Point") -> "Point":
        return Point(self.x + other.x, self.y + other.y)

    def __sub__(self, other: "Point") -> "Point":
        return Point(self.x - other.x, self.y - other.y)

    def __eq__(self, other: "Point") -> bool:
        return self.x == other.x and self.y == other.y

    def __repr__(self) -> str:
        return f"Point<{self.x}, {self.y}>"


class Grid:
    def __init__(self, lines: list[str], diagonal=False):
        self.diagonal = diagonal
        self.grid: list[list[str]] = []
        for line in lines:
            self.grid.append(list(line))

    def get_neighbors(self, p: Point) -> list[Point]:
        adjacents = [Point(x, y) for x, y in [(1, 0), (0, 1), (-1, 0), (0, -1)]]
        diagonals = [Point(x, y) for x, y in [(1, 1), (1, -1), (-1, 1), (-1, -1)]]

        neighbors = [p + n for n in adjacents if self.in_bounds(p + n)]
        if self.diagonal:
            neighbors += [p + n for n in diagonals if self.in_bounds(p + n)]
        return neighbors

    def in_bounds(self, p: Point) -> bool:
        return 0 <= p.y < len(self.grid) and 0 <= p.x < len(self.grid[p.y])

    def get(self, p: Point) -> Optional[str]:
        if self.in_bounds(p):
            return self[p]
        return None

    def __getitem__(self, p: Point) -> str:
        return self.grid[p.y][p.x]

    def __setitem__(self, p: Point, to: str) -> None:
        self.grid[p.y][p.x] = to

    def __iter__(self):
        for y in range(len(self.grid)):
            for x in range(len(self.grid[y])):
                yield Point(x, y)

    def __repr__(self) -> str:
        return "\n".join(["".join(row) for row in self.grid])
